### PR TITLE
EOS-26179: added validations for image versions in upgrade scenario

### PR DESCRIPTION
**Pull request by rahul-tripathi-git**
Signed-off-by: Rahul Tripathi <rahul.tripathi@seagate.com>

# Problem Statement
- Added validations for image versions check

# Design
-  compared the versions present in confstore file and that present in cluste.conf file of the new image .

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

**Review by kailashyadav11 — CHANGES_REQUESTED** [review]
current_full_ver=$(conf_get $CORTX_CONFSTORE "cortx>common>relese>version


please check if the key "release" has a typo error

File: src/setup/cortx_deploy [hunk 1]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
 cluster_id=$(conf_get $(yaml_config $CLUSTER_INFO) "cluster>id")
 log $INFO "Starting provisioning on node $node_id which is part of cluster $cluster_id."
 
+deploy=false
+upgrade=false
+# Check current image version
+CONFSTOR_FILE=$(awk -F:// '{print $2}' <<< $CORTX_CONFSTORE)
+# check if confstore file is empty or not present
+if [ ! -s "$CONFSTOR_FILE" ]; then
```
Comment (tanujashinde0405) [inline src/setup/cortx_deploy:119]: don't mention confstore as file, keep it as CONFSTORE. remove 'file' references for confstore from this script. also, can we validate the location of confstore without splitting the URL?

File: src/setup/cortx_deploy [hunk 2]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    log $INFO "$CORTX_CONFSTORE file not found or file is empty"
+    log $INFO "deployment set to true"
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:121]: This can be debug message...

File: src/setup/cortx_deploy [hunk 3]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
 cluster_id=$(conf_get $(yaml_config $CLUSTER_INFO) "cluster>id")
 log $INFO "Starting provisioning on node $node_id which is part of cluster $cluster_id."
 
+deploy=false
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:114]: Lets not introduce new flags... Ideally we should leverage on using the same flags as mentioned in:
src/provisioner/provisioner.py

`class PROVISIONING_STAGES(Enum):`
`class PROVISIONING_STATUS(Enum):`

File: src/setup/cortx_deploy [hunk 4]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+upgrade=false
+# Check current image version
+CONFSTOR_FILE=$(awk -F:// '{print $2}' <<< $CORTX_CONFSTORE)
+# check if confstore file is empty or not present
+if [ ! -s "$CONFSTOR_FILE" ]; then
+    log $INFO "$CORTX_CONFSTORE file not found or file is empty"
+    log $INFO "deployment set to true"
+    deployment=true
+else
+    current_full_ver=$(conf_get $CORTX_CONFSTORE "cortx>common>release>version")
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:124]: Change variable: `current_image_version`

File: src/setup/cortx_deploy [hunk 5]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    if [ $current_full_ver = "[null]" ]; then
+        deployment=true
+    else
+        # Below can be used when subversioning is considered.
+        #current_primary_ver=$(awk -F. '{print $1 "." $2 "." $3}' <<< $current_full_ver)
+        current_ver=$(awk -F. '{print $4}' <<< $current_full_ver)
+    fi
+fi
+
+image_full_ver=$(sed -e 's/[][]//g' -e 's/"//g' <<< $(conf_get $RELEASE_INFO "VERSION"))
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:135]: Change this to: `update_image_version`

File: src/setup/cortx_deploy [hunk 6]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    deployment=true
+else
+    current_full_ver=$(conf_get $CORTX_CONFSTORE "cortx>common>release>version")
+    # if current version is not available in confstore
+    if [ $current_full_ver = "[null]" ]; then
+        deployment=true
+    else
+        # Below can be used when subversioning is considered.
+        #current_primary_ver=$(awk -F. '{print $1 "." $2 "." $3}' <<< $current_full_ver)
+        current_ver=$(awk -F. '{print $4}' <<< $current_full_ver)
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:131]: Change this to: `current_version`

File: src/setup/cortx_deploy [hunk 7]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    fi
+fi
+
+image_full_ver=$(sed -e 's/[][]//g' -e 's/"//g' <<< $(conf_get $RELEASE_INFO "VERSION"))
+# Below can be used when subversioning is considered.
+#image_primary_ver=$(awk -F- '{print $1}' <<<  $image_full_ver )
+image_ver=$(awk -F- '{print $2}' <<<  $image_full_ver)
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:138]: Change this to: `update_version`

**Review by nitin-seagate — CHANGES_REQUESTED** [review]
Check if this code works for input ConfStore argument is different then default location:
For example: yaml:///var/data/cortx/cluster.conf

File: src/setup/cortx_deploy (hunk 5 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:135]: Done

File: src/setup/cortx_deploy (hunk 6 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:131]: Done

File: src/setup/cortx_deploy (hunk 4 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:124]: Done

File: src/setup/cortx_deploy (hunk 7 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:138]: Done

File: src/setup/cortx_deploy (hunk 3 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:114]: Hi Nitin,
This is a shell script outside the provisioner.py python file.

File: src/setup/cortx_deploy [hunk 8]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+upgrade=false
+# Check current image version
+CONFSTOR_FILE=$(awk -F:// '{print $2}' <<< $CORTX_CONFSTORE)
+# check if confstore file is empty or not present
+if [ ! -s "$CONFSTOR_FILE" ]; then
+    log $INFO "$CORTX_CONFSTORE file not found or file is empty"
+    log $INFO "deployment set to true"
+    deployment=true
+else
+    current_full_ver=$(conf_get $CORTX_CONFSTORE "cortx>common>relese>version")
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

File: src/setup/cortx_deploy [hunk 9]
```diff
@@ -127,9 +168,14 @@ then
     cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c "$CORTX_CONFSTORE" || exit 1
 fi
 
-# Bootstrap Cluster
-log $INFO "cortx_setup cluster bootstrap -c '$CORTX_CONFSTORE' $FORCE_PROVISIONING"
-cortx_setup cluster bootstrap -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+if [[ $deploy = true && $upgrade = false ]]; then
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:171]: Done as per discussion.

File: src/setup/cortx_deploy [hunk 10]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
 
+deploy=false
+upgrade=false
+# Check current image version
+CONFSTOR_FILE=$(awk -F:// '{print $2}' <<< $CORTX_CONFSTORE)
+# check if confstore file is empty or not present
+if [ ! -s "$CONFSTOR_FILE" ]; then
+    log $INFO "$CORTX_CONFSTORE file not found or file is empty"
+    log $INFO "deployment set to true"
+    deployment=true
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:122]: Done

File: src/setup/cortx_deploy [hunk 11]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+else
+    current_full_ver=$(conf_get $CORTX_CONFSTORE "cortx>common>relese>version")
+    # if current version is not available in confstore
+    if [ $current_full_ver = "[null]" ]; then
+        deployment=true
+    else
+        # Below can be used when subversioning is considered.
+        #current_primary_ver=$(awk -F. '{print $1 "." $2 "." $3}' <<< $current_full_ver)
+        current_ver=$(awk -F. '{print $4}' <<< $current_full_ver)
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:131]: Resolving as per discussion.

File: src/setup/cortx_deploy [hunk 12]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    fi
+fi
+
+image_full_ver=$(sed -e 's/[][]//g' -e 's/"//g' <<< $(conf_get $RELEASE_INFO "VERSION"))
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:135]: Resolving as per discussion

File: src/setup/cortx_deploy [hunk 13]
```diff
@@ -110,6 +111,46 @@ node_id=`cat /etc/machine-id`
...
+    fi
+elif [ $current_ver -gt $image_ver ]; then
+    log $ERROR " Current version:$current_full_ver of cortx-all is greater than input image version:$image_full_ver"
+    exit 1
+else
+    deploy=false
+    upgrade=true
+fi
+
 # Apply Cluster Config
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:154]: Done.

File: src/setup/cortx_deploy (hunk 1 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Rmoved the check for consfstore as a file all-together.

File: src/setup/cortx_deploy (hunk 2 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Removed this check as confstore is not to be considered as a file.

File: src/setup/cortx_deploy [hunk 14]
```diff
@@ -58,6 +58,9 @@ function conf_get {
 DEBUG_FLAG=false
 SOLUTION_CONFIG="/etc/cortx/solution"
 CORTX_CONFSTORE="yaml:///etc/cortx/cluster.conf"
+RELEASE_INFO="yaml:///opt/seagate/cortx/RELEASE.INFO"
+DEPLOY=false
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:62]: Lets use DEPLOY_FLAG and UPGRADE_FLAG to be consistent for boolean values

**Review by nitin-seagate — APPROVED** [review]

File: src/setup/cortx_deploy [hunk 15]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
+update_version=$(awk -F- '{print $2}' <<<  $update_image_version)
+
+if [ $current_version -eq $update_version ]; then
+    if [ ! -z $FORCE_PROVISIONING ]; then
+        DEPLOY=true
+    else
+        return 0
+    fi
+elif [ $current_version -gt $update_version ]; then
+    log $ERROR " Current version:$current_image_version of cortx-all is greater than input image version:$update_image_version"
```
Comment (nitin-seagate) [inline src/setup/cortx_deploy:139]: Lets use simple message something like this:
`Cannot update Cortx Image current_version (x.x) > update_version (y.y)`

**Review by nitin-seagate — APPROVED** [review]

File: src/setup/cortx_deploy (hunk 14 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

File: src/setup/cortx_deploy (hunk 15 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:139]: Done

**Review by tanujashinde0405 — APPROVED** [review]

File: src/setup/cortx_deploy [hunk 16]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
-
-# Validate config
-if ! [ -z $VALIDATE_CONF ];
-then
-    log $INFO "cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c '$CORTX_CONFSTORE'"
-    cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c "$CORTX_CONFSTORE" || exit 1
-    log $INFO "cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c '$CORTX_CONFSTORE'"
-    cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c "$CORTX_CONFSTORE" || exit 1
+# Check current image version
+current_image_version=$(conf_get $CORTX_CONFSTORE "cortx>common>release>version")
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:117]: This is not current_image_version. This is current_installed_version. 

Image version is obtained from RELEASE.INFO file.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 17]
```diff
@@ -58,6 +58,9 @@ function conf_get {
 DEBUG_FLAG=false
 SOLUTION_CONFIG="/etc/cortx/solution"
 CORTX_CONFSTORE="yaml:///etc/cortx/cluster.conf"
+RELEASE_INFO="yaml:///opt/seagate/cortx/RELEASE.INFO"
+DEPLOY_FLAG=false
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:62]: Avoid so many flags. This hampers readability of the code. 

We need to determine an action based on facts. So it should be called 
ACTION=""

If condition A; then ACTION=deploy
if condition B; then ACTION=upgrade
and so on

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 18]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
+else
+    # Below can be used when subversioning is considered.
+    #current_primary_ver=$(awk -F. '{print $1 "." $2 "." $3}' <<< $current_image_version)
+    current_version=$(awk -F. '{print $4}' <<< $current_image_version)
 fi
 
-# Bootstrap Cluster
-log $INFO "cortx_setup cluster bootstrap -c '$CORTX_CONFSTORE' $FORCE_PROVISIONING"
-cortx_setup cluster bootstrap -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+update_image_version=$(sed -e 's/[][]//g' -e 's/"//g' <<< $(conf_get $RELEASE_INFO "VERSION"))
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:127]: This is not update_image_version. This is "current_image_version" 

We need to compare current_image_version with current_installed_version.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 19]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
+# Below can be used when subversioning is considered.
+#image_primary_ver=$(awk -F- '{print $1}' <<<  $update_image_version )
+update_version=$(awk -F- '{print $2}' <<<  $update_image_version)
+
+if [ $current_version -eq $update_version ]; then
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:132]: $installed_version -eq $image_version 

For better readability

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 20]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
+    if [ ! -z $FORCE_PROVISIONING ]; then
+        DEPLOY_FLAG=true
+    else
+        return 0
+    fi
+elif [ $current_version -gt $update_version ]; then
+    log $ERROR "Cannot update Cortx Image current_version ($current_image_version) > update_version ($update_image_version)"
+    exit 1
+else
+    DEPLOY_FLAG=false
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:142]: No need to maintain multiple flags as indicated earlier. There will be 5 more commands added in future and this logic will be more complex. Simplify it... 

ACTION=upgrade | deploy | downgrade | replace | recover | etc.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy (hunk 16 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:117]: Done

File: src/setup/cortx_deploy (hunk 20 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:142]: Done.

File: src/setup/cortx_deploy (hunk 17 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:62]: Done

File: src/setup/cortx_deploy (hunk 18 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:127]: Done

File: src/setup/cortx_deploy (hunk 19 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:132]: Done

File: src/setup/cortx_deploy [hunk 21]
```diff
@@ -110,26 +113,60 @@ node_id=`cat /etc/machine-id`
...
+    UPGRADE_FLAG=true
+fi
+
+if [[ $DEPLOY_FLAG = true && $UPGRADE_FLAG = false ]]; then
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

File: src/setup/cortx_deploy (hunk 20 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done.

File: src/setup/cortx_deploy (hunk 19 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:152]: Changed the variable names as mentioned in the comment.

File: src/setup/cortx_deploy [hunk 22]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
-if ! [ -z $VALIDATE_CONF ];
-then
-    log $INFO "cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c '$CORTX_CONFSTORE'"
-    cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c "$CORTX_CONFSTORE" || exit 1
-    log $INFO "cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c '$CORTX_CONFSTORE'"
-    cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c "$CORTX_CONFSTORE" || exit 1
+# Check current image version
+ installed_version=$(conf_get $CORTX_CONFSTORE "cortx>common>release>version")
+# if current version is not available in confstore
+if [ $ installed_version = "[null]" ]; then
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:139]: Remove space after $ 
$installed_version

File: src/setup/cortx_deploy [hunk 23]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
+    current_version=$(awk -F. '{print $4}' <<< $ installed_version)
 fi
 
-# Bootstrap Cluster
-log $INFO "cortx_setup cluster bootstrap -c '$CORTX_CONFSTORE' $FORCE_PROVISIONING"
-cortx_setup cluster bootstrap -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+image_version=$(sed -e 's/[][]//g' -e 's/"//g' <<< $(conf_get $RELEASE_INFO "VERSION"))
+# Below can be used when subversioning is considered.
+#update_prim_version=$(awk -F- '{print $1}' <<<  $image_version )
+update_version=$(awk -F- '{print $2}' <<<  $image_version)
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:150]: We can not call it update version until we know it is update version.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy (hunk 23 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:150]: Done

File: src/setup/cortx_deploy (hunk 22 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:139]: Done

File: src/setup/cortx_deploy [hunk 24]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
+
+if [ $current_version -eq $update_version ]; then
+    if [ ! -z $FORCE_PROVISIONING ]; then
+        ACTION="deploy"
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:154]: action="deploy"

All the variable names must be small letters. Only constants are caps.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 25]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
+    fi
+elif [ $current_version -gt $update_version ]; then
+    log $ERROR "Cannot update Cortx Image current_version ($installed_version) > update_version ($image_version)"
+    exit 1
+else
+    ACTION="upgrade"
+fi
 
+case $ACTION in
+        deploy)
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:166]: indentation is not correct.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy [hunk 26]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
+            fi
+            # Deploy cortx
+            deploy
+            ;;
+        upgrade)
+            # Upgrade cortx
+            upgrade
+            ;;
+        *)
+            log $ERROR "No action selected, check Cortx current_version ($installed_version) and update_version ($image_version)"
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:184]: This error message does not help user in anyway. We need to be more meaningful. 

log $ERROR "Internal error. Could not determine version. Invalid image. Installed version is $current_version, Image version is $image_version."

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/setup/cortx_deploy (hunk 24 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

File: src/setup/cortx_deploy (hunk 25 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done.

File: src/setup/cortx_deploy (hunk 26 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

File: src/setup/cortx_deploy [hunk 27]
```diff
@@ -54,10 +54,33 @@ function conf_get {
...
+    #Upgrade cluster
+    log $INFO "cortx_setup cluster upgrade -c '$CORTX_CONFSTORE' $FORCE_PROVISIONING"
+    cortx_setup cluster upgrade -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+}
 # Constants
 DEBUG_FLAG=false
 SOLUTION_CONFIG="/etc/cortx/solution"
 CORTX_CONFSTORE="yaml:///etc/cortx/cluster.conf"
+RELEASE_INFO="yaml:///opt/seagate/cortx/RELEASE.INFO"
+ACTION="action" # upgrade|deploy|downgrade|replace|recover|etc.
```
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done

**Review by kailashyadav11 — APPROVED** [review]
LGTM

File: src/setup/cortx_deploy [hunk 28]
```diff
@@ -54,10 +54,44 @@ function conf_get {
...
+    cortx_setup cluster bootstrap -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+}
+
+function upgrade {
+    #Upgrade cluster
+    log $INFO "cortx_setup cluster upgrade -c '$CORTX_CONFSTORE' $FORCE_PROVISIONING"
+    cortx_setup cluster upgrade -c "$CORTX_CONFSTORE" $FORCE_PROVISIONING;
+}
+
+function validate{
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:79]: 1. Spacing before {
2. Dont assume global variable. Call this conditionally
```
function validate_conf { 
        log $INFO "cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c '$CORTX_CONFSTORE'"
        cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c "$CORTX_CONFSTORE" || return 1
        log $INFO "cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c '$CORTX_CONFSTORE'"
        cortx_setup config validate -f $(yaml_config $CONFIG_INFO) -c "$CORTX_CONFSTORE" || return 1
        return 0
}

# call 
[ ! -z $VALIDATE_CONF ] && validate_conf
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — APPROVED** [review]

File: src/setup/cortx_deploy [hunk 29]
```diff
@@ -110,27 +144,43 @@ node_id=`cat /etc/machine-id`
...
+# Below can be used when subversioning is considered.
+#update_prim_version=$(awk -F- '{print $1}' <<<  $release_version )
+image_version=$(awk -F- '{print $2}' <<<  $release_version)
+
+[[ $current_version -eq $image_version && -z $FORCE_PROVISIONING ]] &&
+    log $DEBUG "current_version ($installed_version) = image_version ($release_version)" &&
+    return 0
+[[ $current_version -gt $image_version ]] &&
+    log $ERROR "Invalid image. current_version ($installed_version) > image_version ($release_version)" &&
+    exit 1
```
Comment (ujjwalpl) [inline src/setup/cortx_deploy:162]: return 1

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — APPROVED** [review]

File: src/setup/cortx_deploy (hunk 28 above)
Comment (ujjwalpl) [reply src/setup/cortx_deploy:79]: Please change validate_config function to validate_config_file to avoid conflict with this function.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

**Review by ujjwalpl — APPROVED** [review]

File: src/setup/cortx_deploy [hunk 30]
```diff
@@ -110,27 +133,57 @@ node_id=`cat /etc/machine-id`
...
+    ACTION="upgrade"
+fi
 
+case $ACTION in
+        deploy)
+            apply_cluster_config
+            apply_common_config
+            # Validate config
+            if ! [ -z $VALIDATE_CONF ]; then
+                log $INFO "cortx_setup config validate -f $(yaml_config $CLUSTER_INFO) -c '$CORTX_CONFSTORE'"
```
[the remark this answers is not in the document]
Comment (ujjwalpl) [reply src/setup/cortx_deploy]: validate_config should be renamed to validate_config_file

File: src/setup/cortx_deploy (hunk 29 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:162]: Done

File: src/setup/cortx_deploy (hunk 28 above)
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy:79]: Done.

File: src/setup/cortx_deploy (hunk 30 above)
[the remark this answers is not in the document]
Comment (rahul-tripathi-git) [reply src/setup/cortx_deploy]: Done.

**Pull request merged**